### PR TITLE
Improving nxos_ip_interface module

### PR DESCRIPTION
**Pull request by drey**
Added features:

  * Route tags for an IP address at interface level with 'tag' option
  * Support of IPv4 secondary addresses if option `allow_secondary` is
true (false by default). If option `allow_secondary` is false primary
IPv4 address will be replaced

All features support NXAPI and CLI transport, but only with text/raw
output. Currently not possible to get route tags for secondary IPv4/IPv6
addresses from JSON output.

Other changes:
  * Module return state `changed` only when configuration commands were
executed
  * Module result includes multiple prefixes if secondary IPv4 addresses were
found

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ansible/modules/network/nxos/nxos_ip_interface

##### ANSIBLE VERSION
```
ansible --version
ansible 2.2.1.0 (stable-2.2 6176c95838) last updated 2017/02/13 16:48:05 (GMT +300)
  lib/ansible/modules/core: (feature-nxos-ip-interface 2c1f374da0) last updated 2017/02/13 21:58:33 (GMT +300)
  lib/ansible/modules/extras: (detached HEAD 994349b647) last updated 2017/02/13 16:58:44 (GMT +300)
  config file = /Users/idrey/.ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY
I have tested this commit with playbook that can be downloaded by [link](https://gist.github.com/Drey/fa7360f9746c6269f17598b96cb95d82) and Ansible version from branch stable-2.2.

Thanks!

### CODE REVIEW

**Comment by mattclay** [comment]
CI failure due to PEP 8 issues:

```
2017-02-13 19:37:16 ERROR: PEP 8: lib/ansible/modules/network/nxos/nxos_ip_interface.py:164:6: E111 indentation is not a multiple of four (current)
2017-02-13 19:37:16 ERROR: PEP 8: lib/ansible/modules/network/nxos/nxos_ip_interface.py:165:10: E111 indentation is not a multiple of four (current)
2017-02-13 19:37:16 ERROR: PEP 8: lib/ansible/modules/network/nxos/nxos_ip_interface.py:166:6: E111 indentation is not a multiple of four (current)
2017-02-13 19:37:16 ERROR: PEP 8: lib/ansible/modules/network/nxos/nxos_ip_interface.py:167:10: E111 indentation is not a multiple of four (current)
2017-02-13 19:37:16 ERROR: PEP 8: lib/ansible/modules/network/nxos/nxos_ip_interface.py:168:6: E111 indentation is not a multiple of four (current)
2017-02-13 19:37:16 ERROR: PEP 8: lib/ansible/modules/network/nxos/nxos_ip_interface.py:169:10: E111 indentation is not a multiple of four (current)
2017-02-13 19:37:16 ERROR: PEP 8: lib/ansible/modules/network/nxos/nxos_ip_interface.py:541:33: E126 continuation line over-indented for hanging indent (current)
2017-02-13 19:37:16 ERROR: PEP 8: lib/ansible/modules/network/nxos/nxos_ip_interface.py:560:161: E501 line too long (185 > 160 characters) (current)
2017-02-13 19:37:16 ERROR: PEP 8: lib/ansible/modules/network/nxos/nxos_ip_interface.py:627:6: W291 trailing whitespace (current)
2017-02-13 19:37:16 ERROR: PEP 8: lib/ansible/modules/network/nxos/nxos_ip_interface.py:717:13: E126 continuation line over-indented for hanging indent (current)
```

File: lib/ansible/modules/network/nxos/nxos_ip_interface.py [hunk 1]
```diff
@@ -23,7 +23,7 @@
 DOCUMENTATION = '''
 ---
 module: nxos_ip_interface
-version_added: "2.1"
+version_added: "2.2"
```
Comment (privateip) [inline lib/ansible/modules/network/nxos/nxos_ip_interface.py]: `version_added` should not be changed

File: lib/ansible/modules/network/nxos/nxos_ip_interface.py [hunk 2]
```diff
@@ -78,45 +88,66 @@
...
     sample: true
 '''
 
-import json
-import collections
-
 # COMMON CODE FOR MIGRATION
 import re
 
+import ipaddress
```
Comment (privateip) [inline lib/ansible/modules/network/nxos/nxos_ip_interface.py]: `ipaddress` doesn't exist in py2... need to wrap this in a try ... except

**Review by privateip — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/nxos/nxos_ip_interface.py (hunk 2 above)
Comment (drey) [reply lib/ansible/modules/network/nxos/nxos_ip_interface.py]: Fixed the same way as in other modules. Thanks.

File: lib/ansible/modules/network/nxos/nxos_ip_interface.py (hunk 1 above)
Comment (drey) [reply lib/ansible/modules/network/nxos/nxos_ip_interface.py]: Fixed. Thanks!

File: lib/ansible/modules/network/nxos/nxos_ip_interface.py [hunk 3]
```diff
@@ -52,6 +52,16 @@
             - Subnet mask for IPv4 or IPv6 Address in decimal format.
         required: false
         default: null
+    tag:
+        description:
+            - Route tag for IPv4 ot IPv6 Address in integer format.
+        required: false
+        default: 0
```
Comment (gundalow) [inline lib/ansible/modules/network/nxos/nxos_ip_interface.py]: Please add:
`version_added: "2.4"`

File: lib/ansible/modules/network/nxos/nxos_ip_interface.py [hunk 4]
```diff
@@ -52,6 +52,16 @@
...
+    allow_secondary:
+        description:
+            - Allow to configure IPv4 secondary addresses on interface.
+        required: false
```
Comment (gundalow) [inline lib/ansible/modules/network/nxos/nxos_ip_interface.py]: Please add:
`version_added: "2.3"`

File: lib/ansible/modules/network/nxos/nxos_ip_interface.py [hunk 5]
```diff
@@ -78,45 +88,70 @@
...
 
-import json
-import collections
-
 # COMMON CODE FOR MIGRATION
 import re
 
+try:
+    import ipaddress
+    HAS_IPADDRESS = True
```
Comment (gundalow) [inline lib/ansible/modules/network/nxos/nxos_ip_interface.py]: Please docent this in the `requirements:` block
See http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#documentation-block for details.

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/nxos/nxos_ip_interface.py (hunk 5 above)
Comment (drey) [reply lib/ansible/modules/network/nxos/nxos_ip_interface.py]: I did it. Thanks for introduction!

**Review by gundalow — APPROVED** [review]
I'm happy with the docs, not reviewed the code

**Comment by drey** [comment]
Hello, maybe I need to provide some additional information about changes in the commit? Or improve something? Let me know if I can help.

**Comment by drey** [comment]
I have tried to resolve conflicts but could you describe for me how I can keep that module working for both `devel` and `stable-2.2` branches? I think that the only way to do it is a duplication of some code inside nxos_ip_interfaces. Is it the excellent idea?

I have tested all my commits with the branch `stable-2.2` and all features work properly.

**Comment by privateip** [comment]
@Drey in order to merge the PR it must be rebased against the devel branch.

needs_rebase

**Comment by drey** [comment]
@privateip Hi! I have done rebase. Thanks.

**Review by gundalow — APPROVED** [review]

**Comment by drey** [comment]
@privateip Hi! What do you think about my latest commit? Maybe I need to improve something?

**Comment by gundalow** [comment]
@Drey Thanks for this PR (and rebasing it).
We are currently busy working through the final touches for Ansible 2.3.

I've added the Milestone 2.4 marker to this so we will review once 2.3 is complete.

Thanks again

File: lib/ansible/modules/network/nxos/nxos_ip_interface.py [hunk 6]
```diff
@@ -53,12 +52,26 @@
             - Subnet mask for IPv4 or IPv6 Address in decimal format.
         required: false
         default: null
+    tag:
+        description:
+            - Route tag for IPv4 ot IPv6 Address in integer format.
```
Comment (Qalthos) [inline lib/ansible/modules/network/nxos/nxos_ip_interface.py]: IPv4 *or* IPv6, presumably

File: lib/ansible/modules/network/nxos/nxos_ip_interface.py [hunk 7]
```diff
@@ -404,18 +431,42 @@ def validate_params(addr, interface, mask, version, state, intf_type, module):
...
         version=dict(required=False, choices=['v4', 'v6'],
-                         default='v4'),
+                     default='v4'),
         mask=dict(type='str', required=False),
+        tag=dict(required=False, default=0, type='int'),
         state=dict(required=False, default='present',
-                       choices=['present', 'absent']),
+                   choices=['present', 'absent']),
+        allow_secondary=dict(required=False, default=False,
+                             choices=[True, False], type='bool'),
```
Comment (gundalow) [inline lib/ansible/modules/network/nxos/nxos_ip_interface.py]: Please remove choices for `type='bool'`, as `bool` accepts yes,1,true, etc.

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/nxos/nxos_ip_interface.py (hunk 7 above)
Comment (drey) [reply lib/ansible/modules/network/nxos/nxos_ip_interface.py]: Thanks, done

**Pull request merged**